### ISSUE TITLE
FR: Please make dictionary lookup window movable

### ISSUE THREAD

**Issue by mergen3107** [koreader/koreader#7861]
**Does your feature request involve difficulty completing a task? Please describe.**
Problem is that sometimes dictionary lookup window blocks some of the words which are needed to look up. This is with or without the keyboard open.

**Describe the solution you'd like**
Please make this window movable as it is done for many other windows (like dictionary result window itself).

**Describe alternatives you've considered**
Just remembering the word before entry.

**Additional context**
This issue is apparently critical in documents where there are no words to select (scanned pdf/djvu) or the language is unsopported by OCR. See https://github.com/koreader/koreader/issues/7860

Thank you!

**Comment by hius07** [comment]
You can try to add `movable = true,` somewhere here
https://github.com/koreader/koreader/blob/039947886fbc1de89f0236837015b1d817db99f5/frontend/apps/reader/modules/readerdictionary.lua#L684-L687
and look if it works.
Note
https://github.com/koreader/koreader/blob/e4a333a9805a1668860f0a5d134772e21e7904fb/frontend/ui/widget/inputdialog.lua#L162-L166

**Comment by hius07** [comment]
> the InputDialog could 
>  -- be moved under the keyboard, and the user would be locked

Now we can hide the keyboard!

**Comment by mergen3107** [comment]
@hius07 
Thank you! I added `movable = "true",` because without the quotation marks it crashed when I opened look up.
Now it works!

On a second thought I realized, that probably not moving, but making the window transparent might be enough for this purpose.
However, I suppose this only works when the window is allowed to move (I cannot make it transparent without the tweak).

I don't use any gestures apart from along the edges, so I don't know in what ways they might conflict.

**Comment by mergen3107** [comment]
Can I also allow the look up window close when I tap outside it? I know that dictionary results window does that.

**Comment by hius07** [comment]
> Can I also allow the look up window close when I tap outside it?

As simple as movable, no.

**Comment by poire-z** [comment]
> Thank you! I added movable = "true", because without the quotation marks it crashed when I opened look up.

Crash because uimanager.lua expects movable to be the MovableContainer widget:
https://github.com/koreader/koreader/blob/e4a333a9805a1668860f0a5d134772e21e7904fb/frontend/ui/uimanager.lua#L806-L808
while in inputdialog.lua, we named the boolean option "movable" and named the widget we would create if movable=true: "frame", not putting it into self.
So, it would need a small rework of these bits (I checked that there is no other bit of code like that).

But a quick try at moving the dialog under the keyboard, and hiding the keyboard, and trying to bring it back visible did not work.

**Comment by hius07** [comment]
> and hiding the keyboard, and trying to bring it back visible did not work.

If the input is empty, it works.

To reproduce: input something, move the dialog window.
After that the cursor cannot be moved within the input string and the keyboard does not appear after hiding.
In that case, we can touch the screen in the place where the input field was **just before** moving, and the cursor would move and the keyboard would appear.

**Comment by hius07** [comment]
> trying to bring it back visible did not work.

I can make it work with adding

`self:initEventListener()`

after this line
https://github.com/koreader/koreader/blob/c6ccc73d557aa69fd498c77cdc7d800c027cbaff/frontend/ui/widget/inputtext.lua#L495

**Comment by poire-z** [comment]
Dunno.
There's fixing these little cases with hacks (if this is a hack, dunno what you have in mind), like the long-press on down arrow to bring down the keyboard.
Or there is, but it requires much work and thinking, trying to fix this all more generically, like we thought about once in some other issue, like focus a field would bring the keyboard, focus another would keep it, taping outside would unfocus and hide it - until we re tap on a inputtext to focus it again.

**Comment by hius07** [comment]
> dunno what you have in mind

It was understood, that inputtext loses updated position after adding/deleting a char.
So I looked at
https://github.com/koreader/koreader/blob/c6ccc73d557aa69fd498c77cdc7d800c027cbaff/frontend/ui/widget/inputtext.lua#L339-L342
After getting new dimensions
https://github.com/koreader/koreader/blob/c6ccc73d557aa69fd498c77cdc7d800c027cbaff/frontend/ui/widget/inputtext.lua#L495
we reinit EventListener geometry
https://github.com/koreader/koreader/blob/c6ccc73d557aa69fd498c77cdc7d800c027cbaff/frontend/ui/widget/inputtext.lua#L71
and tap (for showing keyboard or moving the cursor inside the input field) and hold (for clipboard popup menu) work again.

**Comment by poire-z** [comment]
Or may be in `InputText:initEventListener()`
`range = function() return self.dimen end,` instead of `range = self.dimen`

I forgot a bit about MovableContainer, and seeing all the commented stuff in `InputText:initEventListener()`, it feels it might need a bit of investigation.
I guess you already made something like this, to avoid the crash in UIManager:
```diff
--- a/frontend/ui/widget/inputdialog.lua
+++ b/frontend/ui/widget/inputdialog.lua
@@ -165,3 +165,3 @@ local InputDialog = InputContainer:new{
     -- be moved under the keyboard, and the user would be locked)
-    movable = false,
+    is_movable = true,

@@ -205,3 +205,3 @@ function InputDialog:init()
     if self.fullscreen then
-        self.movable = false
+        self.is_movable = false
         self.border_size = 0
@@ -443,4 +443,4 @@ function InputDialog:init()
     local frame = self.dialog_frame
-    if self.movable then
-        frame = MovableContainer:new{
+    if self.is_movable then
+        self.movable = MovableContainer:new{
             self.dialog_frame,
@@ -455,3 +455,3 @@ function InputDialog:init()
         },
-        frame
+        self.movable
     }
```
Nothing odd when you hold/pan/swipe when you start in or walk over an InputText?

**Comment by hius07** [comment]
Great! Works perfectly in all inputdialogs everywhere.

**Comment by hius07** [comment]
Besides Text editor and other similar stuff which crashes. Looking at it.

**Comment by hius07** [comment]
./luajit: frontend/ui/widget/container/centercontainer.lua:10: attempt to index a nil value
stack traceback:
	frontend/ui/widget/container/centercontainer.lua:10: in function 'paintTo'
	frontend/ui/widget/container/inputcontainer.lua:85: in function 'paintTo'
	frontend/ui/uimanager.lua:1426: in function '_repaint'
	frontend/ui/uimanager.lua:1486: in function 'forceRePaint'
	frontend/ui/widget/touchmenu.lua:191: in function 'handleEvent'
	frontend/ui/widget/container/inputcontainer.lua:263: in function 'handleEvent'
	frontend/ui/widget/container/widgetcontainer.lua:95: in function 'propagateEvent'
	frontend/ui/widget/container/widgetcontainer.lua:113: in function 'handleEvent'
	frontend/ui/widget/container/widgetcontainer.lua:95: in function 'propagateEvent'
	frontend/ui/widget/container/widgetcontainer.lua:113: in function 'handleEvent'
	frontend/ui/widget/container/widgetcontainer.lua:95: in function 'propagateEvent'
	...
	frontend/ui/widget/container/widgetcontainer.lua:113: in function 'handleEvent'
	frontend/ui/widget/container/widgetcontainer.lua:95: in function 'propagateEvent'
	frontend/ui/widget/container/widgetcontainer.lua:113: in function 'handleEvent'
	frontend/ui/uimanager.lua:1111: in function 'sendEvent'
	frontend/ui/uimanager.lua:53: in function '__default__'
	frontend/ui/uimanager.lua:1573: in function 'handleInputEvent'
	frontend/ui/uimanager.lua:1653: in function 'handleInput'
	frontend/ui/uimanager.lua:1700: in function 'run'
	./reader.lua:323: in main chunk

**Comment by hius07** [comment]
Sorry, needs more testing, a little bit later.

**Comment by hius07** [comment]
> ```diff
> local frame = self.dialog_frame
> ```

Seems we need `self.movable = self.dialog_frame` in line 443 instead.

**Comment by poire-z** [comment]
Better not have/create `self.movable` if it's not movable. May be this then:
```diff
--- a/frontend/ui/widget/inputdialog.lua
+++ b/frontend/ui/widget/inputdialog.lua
@@ -165,3 +165,3 @@ local InputDialog = InputContainer:new{
     -- be moved under the keyboard, and the user would be locked)
-    movable = false,
+    is_movable = true,

@@ -205,3 +205,3 @@ function InputDialog:init()
     if self.fullscreen then
-        self.movable = false
+        self.is_movable = false
         self.border_size = 0
@@ -443,6 +443,7 @@ function InputDialog:init()
     local frame = self.dialog_frame
-    if self.movable then
-        frame = MovableContainer:new{
+    if self.is_movable then
+        self.movable = MovableContainer:new{ -- (UIManager expects this as 'self.movable')
             self.dialog_frame,
         }
+        frame = self.movable
     end
```

**Comment by hius07** [comment]
Yeah, perfectly everywhere.

**Comment by hius07** [comment]
Encountered a little difficulty in fixing.
Let's swipe down the inputdialog to hide it behind the keyboard. Then if I hold the down-arrow to hide the keyboard, the keyboard goes away, and then the moveablecontainer catches the hold_release and restores the inputdialog initial position.
It is because of `self._touch_pre_pan_was_inside` becomes true after the keyboard has gone.
https://github.com/koreader/koreader/blob/2d764d47b858bfaaee2fa19649899553c55ebcc5/frontend/ui/widget/container/movablecontainer.lua#L288-L294
I want the inputdialog window not to restore its initial position after hiding the keyboard.

Tried to change `hold_callback` to `onHoldReleaseKey` in
https://github.com/koreader/koreader/blob/2d764d47b858bfaaee2fa19649899553c55ebcc5/frontend/ui/widget/virtualkeyboard.lua#L128-L137
It actually makes the keyboard to hide not on hold but on release of the down-arrow button, but the moveablecontainer catches the touch_pre_pan and the hold_release anyway.

**Comment by hius07** [comment]
So I try to
-either, stop sending the touch signal after the keyboard has become hidden and I still am holding the screen
-or,  make movablecontainer understand that the touch was not addressed to him (if so, it would not react to the hold_release)

**Comment by poire-z** [comment]
Possibly because once called, in this, self.dimen is zero, so it won't be called?
https://github.com/koreader/koreader/blob/2d764d47b858bfaaee2fa19649899553c55ebcc5/frontend/ui/widget/virtualkeyboard.lua#L213-L218
if yes, may be range should be full screen, so ignore key release can happen, and after this, you might check for it to match the keyboard real dimen?

> It actually makes the keyboard to hide not on hold but on release of the down-arrow button

I don't think I'd like this change on behaviour: it's when I see things happening that I release my finger :)

**Comment by poire-z** [comment]
Other idea: tap anywhere closes keyboard, tap on a text field shows it:
```diff
@@ -455,6 +456,21 @@ function InputDialog:init()
         },
         frame
     }
+    if Device:isTouchDevice() then
+        local GestureRange = require("ui/gesturerange")
+        self.ges_events = {
+            Tap = {
+                GestureRange:new{
+                    ges = "tap",
+                    range = self.dimen,
+                },
+            },
+        }
+    end
+end
+
+function InputDialog:onTap()
+    self._input_widget:onHideKeyboard()
 end

 function InputDialog:getInputText()
```
Keyboard being at top will process a tap on its area. InputText already shows the keyboard on focus, and it being a child of InputDialog, it will see the tap before InputDialog.

**Comment by hius07** [comment]
> Other idea

Brilliant idea! The only change I've made:
`range = self.dimen` to `range = self[1].dimen`, because it closed the keyboard when pressed **between** keys.
Now the keyboard area is excluded, originally:
https://github.com/koreader/koreader/blob/96c895d561a657340621e5ae8b9b537d91af59c5/frontend/ui/widget/inputdialog.lua#L451-L455
Now when the inputdialog is on the screen to hide the keyboard we can touch the screen any point outside the inputfield and the keyboard. Inputdialog buttons still work fine. To show the keaboard back just touch the inputfield.
Works fine when inputdialog is not visible (swiped down behind the keyboard) - touch the screen above the keyboard to hide it.
Works fine in the Multiinputdialog (we can touch the narrow frame to hide the keyboard).
No issues in Text editor (it doesn't work there, as intended).
I'll play with it some time, and if it is OK we can get rid of holding down-arrow to hide the keyboard.

**Comment by hius07** [comment]
Testing is Ok, I see no problems.
I am not happy with the current Clipboard dialog, its keyboard over keyboard is confusing.
I've been actively using Clipboard for a month, and I see no need to have the ability to edit the clipboard content before pasting. I'd paste and then edit if needed.
So I'm going to redesign it from InputDialog to TextViewer, with the same six buttons and functionality.
After that these nice new features can be proposed for merging alltogether: new way of hiding the keyboard, InputDialog movable and transparentable by default, Clipboard dialog without redundant keyboard.

**Comment by poire-z** [comment]
While playing earlier with my above quick hack, I witnessed a few strange things:
- at one point, when moving the dialog under the keyboard, taping to hide keyboard, and I don't remember what: taping outside was then showing the keyboard, taping on the inputtext was hiding it :)
- at another moment, I got 2 input dialogs at the same time (possibly the clipboard thing you mentionned?)

But may be I didn't have some other tweaks/fixes you added.

**Comment by poire-z** [comment]
Closed by #7896.

**Comment by mergen3107** [comment]
Thank you very much guys!
